gpkit/geometric_program.py

```diff
@@ -1,380 +1,99 @@
-                         ["              %s," % constr
-                          for constr in self.constraints] +
-                         ['          ],',
-                          "          substitutions={ %s }," %
-                          pformat(self.substitutions, indent=26)[26:-1],
-                          '          solver="%s")' % self.solver])
-
-    def _latex(self, unused=None):
-        """LaTeX representation of a GeometricProgram.
-        Contains all of its parameters."""
-        return "\n".join(["\\begin{array}[ll]",
-                          "\\text{}",
-                          "\\text{minimize}",
-                          "    & %s \\\\" % self.cost._latex(),
-                          "\\text{subject to}"] +
-                         ["    & %s \\\\" % constr._latex()
-                          for constr in self.constraints] +
-                         ["\\text{substituting}"] +
-                         sorted(["    & %s = %s \\\\" % (var._latex(),
-                                                         latex_num(val))
-                                 for var, val in self.substitutions.items()]) +
-                         ["\\end{array}"])
-
-    def test(self, solver=None, printing=False, skipfailures=False):
-        end = " with default settings."
-        try:
-            self.solve(solver, printing, skipfailures)
-            print(self.__class__.__name__, "solved successfully" + end)
-        except Exception as e:
-            print(self.__class__.__name__, "failed to solve" + end)
-            self.checkbounds()
-            raise(e)
-
-    def solve(self, *args, **kwargs):
-        return self._solve(*args, **kwargs)
-
-    def _solve(self, solver=None, printing=True, skipfailures=False,
-               allownonoptimal=False):
+    def __init__(self, cost, constraints, verbosity=1):
+        self.cost = cost
+        self.constraints = constraints
+        self.posynomials = [cost] + list(constraints)
+        self.cs = np.hstack((mag(p.cs) for p in self.posynomials))
+        if not all(self.cs > 0):
+            raise ValueError("GeometricPrograms cannot contain coefficients"
+                             " less than or equal to zero.")
+        self.exps = functools_reduce(add, (x.exps for x in self.posynomials))
+        self.varlocs, self.varkeys = locate_vars(self.exps)
+        # k [j]: number of monomials (columns of F) present in each constraint
+        self.k = [len(p.cs) for p in self.posynomials]
+        # p_idxs [i]: posynomial index of each monomial
+        p_idx = 0
+        p_idxs = []
+        for p_len in self.k:
+            p_idxs += [p_idx]*p_len
+            p_idx += 1
+        self.p_idxs = np.array(p_idxs)
+
+        self.A, missingbounds = genA(self.exps, self.varlocs)
+
+        if verbosity > 0:
+            for var, bound in sorted(missingbounds.items()):
+                print("%s has no %s bound" % (var, bound))
```

Comment:
Why not make verbosity boolean if it's used in this manner? Just trying to maintain the same kwarg names for user-friendliness? (that does seem valid...)